File: app/dtos/advices.py
from datetime import date, datetime
from enum import StrEnum
from typing import Annotated, Literal

from pydantic import BaseModel, Field


class AdviceTriggerType(StrEnum):
    AUTO = "AUTO"
    MANUAL = "MANUAL"


class AdviceFeedbackType(StrEnum):
    HELPFUL = "HELPFUL"
    NOT_HELPFUL = "NOT_HELPFUL"


class AdviceGenerateRequest(BaseModel):
    trigger_type: AdviceTriggerType = AdviceTriggerType.MANUAL


class DailyAdviceResponse(BaseModel):
    advice_id: int
    advice_date: date
    title: str
    advice_text: str
    provider: str
    model_name: str
    trigger_type: AdviceTriggerType
    generated: bool
    created_at: datetime
    source_type: Literal["RULE_BASED", "LLM"] = "RULE_BASED"
    remaining_regeneration_count: int = 2


class AdviceHistoryItemResponse(BaseModel):
    advice_id: int
    advice_date: date
    title: str
    advice_text: str
    trigger_type: AdviceTriggerType
    source_type: Literal["RULE_BASED", "LLM"] = "RULE_BASED"
    feedback_type: AdviceFeedbackType | None = None
    created_at: datetime


class AdviceFeedbackCreateRequest(BaseModel):
    feedback_type: AdviceFeedbackType
    comment: Annotated[str | None, Field(default=None, max_length=500)]


class AdviceFeedbackCreateResponse(BaseModel):
    feedback_id: int
    advice_id: int
    feedback_type: AdviceFeedbackType
    created_at: datetime

File: frontend/src/api/advices.ts
import { apiRequest } from "./client";

export type DailyAdvice = {
  advice_id: number;
  advice_date: string;
  title: string;
  advice_text: string;
  provider: string;
  model_name: string;
  trigger_type: "AUTO" | "MANUAL";
  generated: boolean;
  created_at: string;
  source_type: "RULE_BASED" | "LLM";
  remaining_regeneration_count: number;
};

export type AdviceHistorySort = "LATEST" | "OLDEST";

export type AdviceHistoryItem = {
  advice_id: number;
  advice_date: string;
  title: string;
  advice_text: string;
  trigger_type: "AUTO" | "MANUAL";
  source_type: "RULE_BASED" | "LLM";
  feedback_type: "HELPFUL" | "NOT_HELPFUL" | null;
  created_at: string;
};

export type AdviceFeedbackPayload = {
  feedback_type: "HELPFUL" | "NOT_HELPFUL";
  comment?: string | null;
};

export type AdviceFeedbackResponse = {
  feedback_id: number;
  advice_id: number;
  feedback_type: "HELPFUL" | "NOT_HELPFUL";
  created_at: string;
};

export function getTodayAdvice(token?: string) {
  return apiRequest<{ data: DailyAdvice }>("/daily-advices/today", { token });
}

export function getAdviceHistory(sort: AdviceHistorySort = "LATEST", token?: string) {
  return apiRequest<{ data: AdviceHistoryItem[] }>(`/daily-advices/history?sort=${sort}`, { token });
}

export function generateTodayAdvice(token?: string) {
  return apiRequest<{ data: DailyAdvice }>("/daily-advices/generate", {
    method: "POST",
    body: JSON.stringify({ trigger_type: "MANUAL" }),
    token,
  });
}

export function createAdviceFeedback(adviceId: number, payload: AdviceFeedbackPayload, token?: string) {
  return apiRequest<{ data: AdviceFeedbackResponse }>(`/daily-advices/${adviceId}/feedbacks`, {
    method: "POST",
    body: JSON.stringify(payload),
    token,
  });
}
